### ISSUE TITLE
Unable to load the pb file from Custom training model

### ISSUE THREAD

**Issue by Kabster17** [szaza/android-yolo-v2#30]
Hi Szaza, 

Thanks a lot for your contribution. I am using Windows 10 and I have incorporated training of my custom model using tiny-yolo to identify 2 classes. 
Now, I would like to load it into an android app however I am unable to use the app on my phone, it keeps crashing. 
Is it because I did my training of my custom model using tensorflow 1.14.0? 
As i am new to this area of Android Studio and applications. I would really appreciate some guidance. Thanks.

**Comment by ZZANZU** [comment]
did you actually uploaded .pb file into your asset folder?

**Comment by Kabster17** [comment]
Hi ZZANZU, 
Yes, i did upload the .pb file into the asset folder but i dont get it why its not loading the model. I used tensorflow 1.14 gpu to train, not sure if there is a possible mismatch between the tensorflow dependencies and with my trained model. Could you share your experience if you uploaded your own custom model?

**Comment by szaza** [comment]
Hi @Kabster17,
It can be because of a mismatch between the tensorflow versions, so if you use tensorflow 1.14.0 for training and 1.6.0 on Android side, than you can have problems. Did you try to increase TF version in the mobile app from 'org.tensorflow:tensorflow-android:1.6.0'? As I see the highest version is 1.13.1, however it is still not 1.14.0, so I'm not sure if it will work for you. https://mvnrepository.com/artifact/org.tensorflow/tensorflow-android

As an alternative solution you can try to downgrade the TF version used for training to 1.6.0, or just to use the same version for training and on the mobile app.
I hope using the same version on both sides will solve your problem. If you still have problems please let me know an we will figure out something.

Best regards,
  Zoltan

**Comment by Kabster17** [comment]
Hi Zoltan, 

Thanks a lot for your reply man. Really appreciate it. Yeah i found out that i needed to downgrade to tf  to 1.6.0 for the training and did some training with some dummy data for about 50 epochs after which I did manage to install/run the app without it crashing and no errors. However, it is unable to detect anything yet hopefully more training and lower loss rate would solve that.  
I have a few questions regarding your application:
1. Is it possible to set a threshold value for the apk/app with regards to the detection of yolo model?
2. Have you ventured into the incorporating on image capturing and video recording as possible enhancements to your app?

**Comment by szaza** [comment]
Hi @Kabster17, 
yes, usually 50 epochs are not enough, more training is needed.

1. yes it is possible in the classifier class itself: https://github.com/szaza/android-yolo-v2/blob/master/src/org/tensorflow/yolo/YOLOClassifier.java; You have to change the value of the THRESHOLD constant.
2. I did not considered video recording and image capturing yet, however it would be a great feature. Would you like to work on it? I would appreciate it.

**Comment by Kabster17** [comment]
Hi Zoltan @szaza , 

Thank you for your reply. Yes, i am looking into ways to enhance your application to perform image capturing and video recording with the detection bounding boxes being present. I think it would make the app more complete. 
But I am really new to android studio and JAVA. Do you have any resources to recommend/suggest to have a better understanding of your app and help me kick start feature enhancements? Thanks.

**Comment by szaza** [comment]
Hi,
thanks for further development, I'm really glad when somebody adds values to the application.
I can only recommend you the official Android development guides + stack overflow: https://developer.android.com/guide. In my opinion that is the best combination.
Best regards,
   Zoltan

**Comment by Kabster17** [comment]
Hi @szaza , 

Yeah sure. Thanks for your help. 
I have an issue now, after training my model consisting of dataset of 2 images and 3 classes. I have brought the loss values to about 0.15. 
Yet, i am unable to detect using the app after uploading my custom model.pb file into the assets folder. 
I have changed the Threshold to a constant as shown:
private final static double THRESHOLD = 0.001;
from 
private final static float THRESHOLD = 0.3f; 

Is there something wrong with my edit or method of training? I am using the same images i used to train to test the app's detection with.

**Comment by szaza** [comment]
Hi, 
I have no idea what could be wrong, however I've created a guideline for training YOLO. Could you please try to follow step-by-step this guideline, just to make sure that your installation works properly: https://sites.google.com/view/tensorflow-example-java-api/complete-guide-to-train-yolo?
Have you tested your generated pb file on desktop side? Does it recognize your images on the desktop environment?

**Comment by Kabster17** [comment]
Hi @szaza, 

I have tested my generated pb file on desktop side, and it gives me no detection as well. It does not recognize the images. Like i have mentioned before, I have trained using Tiny-Yolo using tensorflow 1.14.0 version. It is a set of custom data but i dont understand using the pb files and meta files to perform detection is not occuring. Have you experienced this? 

I trained for 5 classes using about 100 images per class to train for about 25000 steps but after generating the pb files using this command:
->python flow --model cfg/tiny-yolo-voc-5c.cfg --load -1 --savepb 
and i used this command to test the generated pb file:
->python flow --pbLoad built_graph/tiny-yolo-voc-5c.pb --metaLoad built_graph/tiny-yolo-voc-5c.meta --imgdir sample_img/ --threshold 0.01 --gpu 1.0

Using the pb files, i am not able to get any detection but loading the images thru the checkpoint model, I am able to achieve good detection. 
I am not sure why the pb files generated cannot give me any detection. Do you have any idea?

**Comment by szaza** [comment]
Hi,
I had similar behavior when I trained the model by using DarkNet and I tried to export the .pb file using DarkFlow. In my opinion the .pb generation is wrong. As I remember there was a darkflow bug.
I wrote a topic about it: https://sites.google.com/view/tensorflow-example-java-api/complete-guide-to-train-yolo/convert-darknet-weights-to-pb-file.
I'm not sure if it helps you, but it worth a try.

**Comment by Kabster17** [comment]
Hi @szaza,

I just started to read up on the darkflow bug with regards to the generation of the .pb files. Let me just check with you the following:
1. I trained my custom model with yolov2 weights instead of tiny-yolo-voc weights, is there a difference between them? I mean I am able to detect using the normal python flow method as mentioned earlier. 
2. What is the issue with darkflow being unable to generate .pb files after 14000 steps in training? Is it true and whats the work around it? 
I have mentioned the link below that directs you to link:
https://github.com/thtrieu/darkflow/issues/698

**Comment by szaza** [comment]
Hi Kabster17,

"I trained my custom model with yolov2 weights instead of tiny-yolo-voc weights, is there a difference between them? I mean I am able to detect using the normal python flow method as mentioned earlier." - yes, that can be a problem. YoloV2 is a larger model than tiny-yolo, so easily can happen that you don't have enough memory in your mobile phone to execute it.

Is your installation able to work with the sample weights, that can be found here: https://github.com/szaza/android-yolo-v2/tree/master/assets?

**Comment by Kabster17** [comment]
Hi @szaza,

Yes i could detect and work with your sample .pb and labels.txt files. 
Actually, i have gotten the app working and all but it is kinda detecting rubbish with the threshold being set is at 0.001 or no detection at all.
I trained a model to detect 2 classes with about 100 images in total and the 2 classes are mostly present together in an image. The loss rate went down to 1.03 at 14000 steps and the tiny-yolo model could detect using both the .pb file @threshold=0.01 and with the checkpoint file with the same threshold value of 0.01.
But when i upload it to the assets folder to install the app, i get very poor detection results.
Can i know at what loss rate and steps did you train your model until and does the frame dimensions affect the detection results?
I tested the app to detect the exact images/dataset that i used to train the model but yet i am not unable to get any detection/bounding boxes.

**Comment by Kabster17** [comment]
> Hi @szaza,
> 
> Yes i could detect and work with your sample .pb and labels.txt files.
> Actually, i have gotten the app working and all but it is kinda detecting rubbish with the threshold being set is at 0.001 or no detection at all.
> I trained a model to detect 2 classes with about 100 images in total and the 2 classes are mostly present together in an image. The loss rate went down to 1.03 at 14000 steps and the tiny-yolo model could detect using both the .pb file @Threshold=0.01 and with the checkpoint file with the same threshold value of 0.01.
> But when i upload it to the assets folder to install the app, i get very poor detection results.
> Can i know at what loss rate and steps did you train your model until and does the frame dimensions affect the detection results?
> I tested the app to detect the exact images/dataset that i used to train the model but yet i am not unable to get any detection/bounding boxes.

Can you suggest to me what can i do to improve the detection results with regards to the training and what not. As I would like to get some sort of decent detection results to look further into improving the model or evaluating the model. Thanks.

**Comment by szaza** [comment]
Sorry for the late reply, I was very busy.
Did you manage to detect objects on images in the meantime?
